pkg/continuoustest/manager.go

```diff
@@ -71,8 +73,12 @@ func (m *Manager) Run(ctx context.Context) error {
 			if m.cfg.SmokeTest {
 				if err != nil {
 					level.Info(m.logger).Log("msg", "Test failed", "test", t.Name(), "err", err)
+					util_log.Flush()
```

Comment:
We definitely shouldn't be calling `os.Exit()` from library code or using the global logger. I think the existing code is correct: it returns unconditionally inside the `if m.cfg.SmokeTest {` block.

Reply:
I agree it's not nice. However, services by their nature don't really exit conditionally like the smokes in continuous-test. It should be handled by the manager. I was contemplating putting this logic in the module init which is the correct layer imo but also a bit of a slap on what modules and services are supposed to behave like. Thoughts?

Reply:
Then again I think perhaps the best layer to handle this might be the dockerfile usage of the module.

Reply:
I think we can make this work. It looks like the overall `ServiceManager` Mimir uses runs until either a module fails (returns an error from its run method) or the server gets a signal. 

There's an error in dskit `modules.ErrStopProcess` that is used for one-shot tasks to indicate that Mimir should shutdown after the service returns even when there's not a real "error" that's happened. There's an [example of that in the flusher](https://github.com/grafana/mimir/blob/1e903e4620504a265d49ccee621f09e17b3a18f8/pkg/flusher/flusher.go#L92-L94) module.

Reply:
You might need to rework this code a bit to get it working since `group.Go()` cancels the group after the first non-nil error so you can't return `modules.ErrStopProcess` directly from the goroutine for each test. You'd have to check `m.cfg.SmokeTest` again before returning the error from `group.Wait()`